Expand on best practice areas (#128)
cncf/cnf-wg: doc/best_cnf_dev.md

# Best Practices for CNF Developers

## Table of Contents
* [1.0 Installation and Upgrade](#1.0)
* [2.0 Configuration and Lifecycle](#2.0)
* [3.0 Hardware support](#3.0)

> * [4.0 Microservice](#4.0)

Author: ```suggestion
* [4.0 Microservices](#4.0)
```

* [5.0 Compatiblity](#5.0)
* [6.0 State](#6.0)
* [7.0 Security](#7.0)
* [8.0 Scaling](#8.0)
* [9.0 Observability and Diagnostics](#9.0)

<a name="1.0"></a>

> ## 2.0 Installation and Upgrade

Reviewer: ```suggestion
## 1.0 Installation and Upgrade
```

<!-- markdown-link-check-disable-next-line -->
**Link to CCDP**: [CCDP-0002](../ccdps/xyz.md)

<a name="3.0"></a>
## 3.0 Hardware support

How to co-ordinate hardware to best enable CNF efficiency.  Low-level access to hardware such as NICs; access to acceleration technologies; efficient management of standard resources, including manipulation of lower-level system components such as CPU cores, CPU cache and memory.

<a name="4.0"></a>
## 4.0 Microservices

The best way to use microservice-based design patterns to deliver CNF functionality.

<a name="5.0"></a>
## 5.0 Compatiblity

How to ensure CNFs work on as many platforms, in as many places, as possible; how to ensure that cloud platforms enable as many CNFs as possible; how to make CNFs work with each other and common elements in the operational stack.

<a name="6.0"></a>
## 6.0 State

> Management of short-lived and long lived state, including state associated with network flows, configuration data, billing records and other data according to the varying requirements of resiliency, rate of change, access rate and persistence.

Reviewer 2: Can 'billing records' be generalised to something like 'subscriber activity data'

Author: Yes, certainly.  Can you propose a change I can accept? (The doc icon with '+-' on it in the github comment toolbar)

Reviewer 2: OK, I will try to find the button and then submit a PR on your PR (or whatever the correct term is)

Author: It should just be a change as a part of a comment on this pull request.

<img width="789" alt="Screenshot 2021-04-09 at 09 13 03" src="https://user-images.githubusercontent.com/1719713/114209807-e6c48800-9913-11eb-835a-0ede61dafd4f.png">

(See the icon above the word 'comment' in the screenshot?  If not, we may need to add a permission bit for you.)

Author: (If this is a problem don't spend much time on it; just tell me it's all screwed up for you and I'll propose the fix.)

Reviewer 2: ```suggestion
Management of short-lived and long lived state, including state associated with network flows, configuration data, subscriber activity data and other data according to the varying requirements of resiliency, rate of change, access rate and persistence.
```

Reviewer 2: @iawells I think I did the right thing, let me know if not...

Author: Perfect, thanks.  I'll talk to @taylor and @jeffsaelens about adding this to the reviewer help.

<a name="7.0"></a>
## 7.0 Security

How to ensure that components are protected against security issues, including security advisories on software components, defence against attacks, and defence in depth.

<a name="8.0"></a>
## 8.0 Scaling

Running of CNFs at a variety of different scales to manage different traffic requirements and subscriber counts.  Changing the scale of CNFs.

<a name="9.0"></a>
## 9.0 Observability and Diagnostics

> How to get critical data about when things are going wrong and how to determine what must be done to put them right.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, and to drill down into that area to determine a root cause and a fix.

Reviewer 3: Is it fair to aspire for most of this to be automated?  If so, maybe it can read:

How to get critical data about when things are going wrong and to determine what must be done to put them right with limited or no human intervention.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, determine root cause, and correct.

Author: I see your point and I agree.  I might reword it to make it more that this is consumable by an automated system or a human being than prioritise automation over humans.  We have to serve both audiences anyway.

I don't disagree with you that automation is actually more important, but the humans involved sometimes do :)

Author: ```suggestion
How to get critical data about when things are going wrong and how to determine what must be done to put them right.  The detection and correction may be through the actions of an operator or via an automated system.  Using logs and metrics from all components in the system to narrow down the area where a problem exists, and to drill down into that area to determine a root cause and a fix.
```

Reviewer 3: Thanks, this looks good to me.
